src/v2/guide/composition.md

```diff
@@ -0,0 +1,529 @@
+
+``` html
+<error-note type="success">
+  <span slot="bold">Welcome!</span>
+  Would you like to take the tour?
+</error-note>
+```
+
+Above, a developer has hijacked the component for a new kind of note that is not actually an error. The wrapper version of this component more strictly controls the API, so does not allow this kind of abuse.
+
+<p class="tip">In the example above, <code>type="success"</code> <em>would</em> actually affect the wrapped <code>AppNote</code> component, but as an attribute rather than a prop. To prevent attributes from automatically being added to the root element of a component, use the <a href="../api/#inheritAttrs"><code>inheritAttrs: false</code></a> option.</p>
+
+
+
+## Factory Functions
+
+[**Factory functions**](https://medium.com/javascript-scene/javascript-factory-functions-with-es6-4d224591a8b1) are functions that return an object - in our case, component options:
+
+``` js
+function createComponent (options) {
+  return {
+    // ... dynamically generated component config ...
+  }
+}
+```
+
+> If you're familiar with React, their [higher-order components](https://reactjs.org/docs/higher-order-components.html) are one kind of factory function.
+
+Factory functions allow you to better control and simplify how component variations are created. For example, it could be argued that the `ErrorNote` we [created earlier with extension](#An-Example-Use-Case) would be conceptually simpler as a factory function:
+
+``` js
+function createNote (options) {
+  options = options || {}
+  options.props = options.props || {}
+  options.slots = options.slots || {}
+
+  return {
+    props: {
+      type: {
+        type: String,
+        default: options.props.type || 'info'
+      }
+    },
+    template: `
+      <p :class="['note', type]">
+        <strong>
+          <slot name="bold">${
+            options.slots.bold || ''
+          }</slot>
+        </strong>
+        <slot>${
+          options.slots.default || ''
+        }</slot>
+      </p>
+    `
+  }
+}
+```
+
+Then we could create our `AppNote` and `ErrorNote` components with:
+
+``` js
+var AppNote = createNote()
+
+var ErrorNote = createNote({
+  props: {
+    type: 'error'
+  },
+  slots: {
+    bold: 'Error!'
+  }
+})
+```
+
+Using this strategy, all notes in the app maintain unified behavior, but _you_ control the API for creating them. This can be especially useful when you want to:
+
+- Create many variations of a component, abstracting away the specific implementation details.
+- Give designers, less experienced developers, or members of a very large team the ability to create new component variations, but strictly control what they're allowed to change.
+- Dynamically build template strings at runtime, as a more familiar alternative to render functions.
+
+We'll demonstrate the last advantage in the next section.
+
+### Computed Components
```

Comment:
Not really a fan of computed components - creating component definitions on the fly is always inefficient, especially when it involves dynamically generating template strings. I'd just suggest people stick to render functions and remove this section.

Reply:
I disagree. This is one of the most asked questions on the forum, in a sense.

there seem to be many people out there who are stuck with backends that send string templates to the frontend, and they all end up trying to insert them into their apps with v-html and wonder why the template syntax is not being processed.

The solution for these people is to create a dynamic component from the template string that they received.

So I strongly vote for keeping this in the guide, and just explain the downsides of that approach. That would give me a piece of documentation that I can refer these people to.

Reply:
I will look one up tomorrow.

Reply:
This is one, but maybe not the best one:

https://forum.vuejs.org/t/innerhtml-compilation-vue-2/8780

I'll dig for some more.

Reply:
https://forum.vuejs.org/t/dynamically-compile-router-link/7410

Reply:
Thanks @LinusBorg , The "dynamic router links" one is probably the example that mostly drives home the point. I would probably still try to direct users in the direction of using "funcitonal" components for this kind of behavior, yet if we are going to add this, I would recommend using the router example as it's more clear for what it tries to achieve.